### PR TITLE
Manage database resilience and add configuration_database in /status

### PR DESCRIPTION
**Pull request by kadhikari**
* Jormungandr in navitia connects to the database to validate user with or without token and uses configurations saved in the database via navitia-configurator. 
* For an unknown reason if the database is not available  or accessible, navitia should work without any connection to the database and using default configuration values.

* Api /status with or without coverage should return a response with information on one or all the coverages.

Tickets:
https://jira.kisio.org/browse/NAVITIAII-3565
https://jira.kisio.org/browse/NAVITIAII-3471

**Note: Tested in local with two coverages**
* Database available / not available
* Coverages available: 0, 1 and 2... 

Below are some screen shots:
Status without coverage and database connected:
connected_status_sans_coverage

Status with coverage and database connected:
connected_coverage_avec_status

Status with coverage and database not available:
no_SQL_connection_coverage_avec_status

Coverage without status and database not available:
no_SQL_connection_coverage_sans_status

Status without coverage and database not available:
no_SQL_connection_status_sans_coverage

### CODE REVIEW

File: source/jormungandr/jormungandr/authentication.py [hunk 1]
```diff
@@ -127,6 +128,8 @@ def has_access(region, api, abort, user):
     if current_app.config.get('PUBLIC', False):
         # if jormungandr is on public mode we skip the authentification process
         return True
+    if not can_connect_to_database:
```
Comment (xlqian) [inline source/jormungandr/jormungandr/authentication.py]:
```suggestion
    if not can_connect_to_database():
```

Comment (xlqian) [reply source/jormungandr/jormungandr/authentication.py]: :smile:

File: source/jormungandr/jormungandr/instance_manager.py [hunk 2]
```diff
@@ -266,6 +267,8 @@ def stop(self):
             self.thread_event.set()
 
     def _filter_authorized_instances(self, instances, api):
+        if not can_connect_to_database():
+            return instances
```
Comment (xlqian) [inline source/jormungandr/jormungandr/instance_manager.py]: what if `instances` is empty?

File: source/jormungandr/jormungandr/modules/v1_routing/resources/Index.py [hunk 3]
```diff
@@ -109,4 +110,9 @@ def get(self):
         if callable(cache_status_op):
             response['redis'] = cache_status_op()
 
+        if can_connect_to_database():
+            response['configuration_database'] = 'connected'
+        else:
+            response['configuration_database'] = 'no_SQL_connection'
```
Comment (xlqian) [inline source/jormungandr/jormungandr/modules/v1_routing/resources/Index.py]:
```suggestion
        response['configuration_database'] = 'connected' if can_connect_to_database() else 'no_SQL_connection'
```

File: source/jormungandr/jormungandr/interfaces/v1/Status.py [hunk 4]
```diff
@@ -54,4 +55,9 @@ def get(self, region=None, lon=None, lat=None):
             region_str
         )
         logging.getLogger(__name__).info("status reponse: %s", response)
+        if can_connect_to_database():
+            response['status']['configuration_database'] = 'connected'
+        else:
+            response['status']['configuration_database'] = 'no_SQL_connection'
```
Comment (xlqian) [inline source/jormungandr/jormungandr/interfaces/v1/Status.py]:
```suggestion
        response['configuration_database'] = 'connected' if can_connect_to_database() else 'no_SQL_connection'
```

File: source/jormungandr/jormungandr/travelers_profile.py [hunk 5]
```diff
@@ -124,6 +125,8 @@ def make_traveler_profile(cls, coverage, traveler_type):
         """
         if app.config[str('DISABLE_DATABASE')]:
             return default_traveler_profiles[traveler_type]
+        if not can_connect_to_database():
+            return default_traveler_profiles[traveler_type]
```
Comment (xlqian) [inline source/jormungandr/jormungandr/travelers_profile.py]:
```suggestion
        if app.config[str('DISABLE_DATABASE')] or not can_connect_to_database():
            return default_traveler_profiles[traveler_type]
```

File: source/jormungandr/jormungandr/utils.py [hunk 6]
```diff
@@ -671,3 +671,20 @@ def has_invalid_reponse_code(objects):
...
+
+def can_connect_to_database():
+    if hasattr(g, 'can_connect_to_database'):
+        return g.can_connect_to_database
+    try:
+        engine = models.db.engine
+        connection = engine.connect()
+        connection.close()
+        g.can_connect_to_database = True
+        print('++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++')
```
Comment (xlqian) [inline source/jormungandr/jormungandr/utils.py]: ???????????????????????????????????????

File: source/jormungandr/jormungandr/utils.py [hunk 7]
```diff
@@ -671,3 +671,20 @@ def has_invalid_reponse_code(objects):
...
+    except Exception:
+        g.can_connect_to_database = False
+        print('--------------------------------------------------------------------------------------------')
```
Comment (xlqian) [inline source/jormungandr/jormungandr/utils.py]: ??????????????????????????????????????????????????

File: source/jormungandr/jormungandr/interfaces/v1/Status.py (hunk 4 above)
Comment (kadhikari) [reply source/jormungandr/jormungandr/interfaces/v1/Status.py]: :+1:

File: source/jormungandr/jormungandr/instance_manager.py (hunk 2 above)
Comment (kadhikari) [reply source/jormungandr/jormungandr/instance_manager.py]: :+1:

**Review by pbench — APPROVED** [review]

File: source/jormungandr/jormungandr/external_services/external_service_provider_manager.py [hunk 8]
```diff
@@ -33,11 +33,12 @@
 import datetime
 from jormungandr import utils
 from jormungandr.exceptions import ConfigException
+from jormungandr.utils import can_connect_to_database
 
 
 class ExternalServiceManager(object):
     def __init__(
-        self, instance, external_service_configuration, external_service_getter=None, update_interval=600
```
Comment (benoit-bst) [inline source/jormungandr/jormungandr/external_services/external_service_provider_manager.py]: this is a occasion to set up **update_interval** with a new val ?

File: source/jormungandr/jormungandr/instance.py [hunk 9]
```diff
@@ -214,7 +215,7 @@ def get_external_service_providers_from_db(self):
         """
         :return: a callable query of external services associated to the current instance in db
         """
-        result = self._get_models().external_services
+        result = self._get_models().external_services or None
```
Comment (benoit-bst) [inline source/jormungandr/jormungandr/instance.py]: or None ?

**Review by benoit-bst — APPROVED** [review]

File: source/jormungandr/jormungandr/instance.py (hunk 9 above)
Comment (kadhikari) [reply source/jormungandr/jormungandr/instance.py]: A correction is on the way

File: source/jormungandr/jormungandr/external_services/external_service_provider_manager.py (hunk 8 above)
Comment (kadhikari) [reply source/jormungandr/jormungandr/external_services/external_service_provider_manager.py]: :+1:

**Review by benoit-bst — APPROVED** [review]

**Review by xlqian — APPROVED** [review]

**Review by benoit-bst — APPROVED** [review]

**Pull request merged**